Sources/LeafKit/LeafLexer.swift

```diff
@@ -132,123 +141,191 @@ struct LeafLexer {
         
-        switch state {
-        case .normal:
-            switch next {
-            case .backSlash:
-                // consume '\' only in event of '\#'
-                // otherwise allow it to remain for other
-                // escapes, a la javascript
-                if src.peek(aheadBy: 1) == .tagIndicator {
-                    src.pop()
-                }
-                // either way, add raw '#' or '\' to registry
-                return src.pop().flatMap { .raw(.init($0)) }
-            case .tagIndicator:
-                // consume `#`
-                src.pop()
-
-                // if tag indicator is followed by an invalid token, assume that it is unrelated to leaf
-                  guard let next = src.peek() else { throw "can not terminate document with tag indicator" }
-                  if next.isValidInTagName || next == .leftParenthesis {
-                      state = .tag
-                      return .tagIndicator
-                  } else {
-                      state = .normal
-                      return .raw(Character.tagIndicator.description)
-                  }
-            default:
-                // read until next event
-                let slice = src.readWhile { $0 != .tagIndicator && $0 != .backSlash } ?? ""
-                return .raw(slice)
-            }
-        case .tag:
-            switch next {
-            case .leftParenthesis:
-                // '#('
-                state = .parameters(depth: 0)
-                return .tag(name: "")
-            case let x where x.isValidInTagName:
-                // collect the named tag, letters only
-                let val = src.readWhile { $0.isValidInTagName }
-                guard let name = val else { fatalError("switch case should disallow this") }
-                
-                let trailing = src.peek()
-                if trailing == .colon { state = .body }
-                else if trailing == .leftParenthesis { state = .parameters(depth: 0) }
-                else { state = .normal }
-                
-                return .tag(name: name)
+        switch   (state,       isTagID, isTagVal, isCol, next) {
+            case (.raw,        false,   _,        _,     _):     return lexRaw()
+            case (.raw,        true,    _,        _,     .some): return lexCheckTagIndicator()
+            case (.tag,        _,       true,     _,     _):     return lexNamedTag()
+            case (.tag,        _,       false,    _,     _):     return lexAnonymousTag()
+            case (.parameters, _,   _,   _,  _):                 return try lexParameters()
+            case (.body,       _,   _, true,  _):                return lexBodyIndicator()
+            /// Ambiguous case  - `#endTagName#` at EOF. Should this result in `tag(tagName),raw(#)`?
+            case (.raw,        true,    _,        _,     .none):
+            throw LexerError(.internalError("Unescaped # at EOF"), src: src, lexed: lexed)
```

Comment:
If anyone thinks this is ugly, I don't entirely disagree but it makes it MUCH easier to read & maintain at the moment in my opinion. The switch tuples can be simplified in the future but I think this helps break down where unexpected failures can happen.

Reply:
🙃🙃🙃